---
title: "Associations"
author: "Carlos Ruiz"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
```

# Introduction

In this report, we will show the different associations performed.

```{r}
library(qqman)
library(kableExtra)
library(tidyverse)
library(GenomicRanges)
library(gwascat)
library(SNPassoc)
library(epiDisplay)
```

# GWAS

First, we run a GWAS analysis with the imputed SNPs passing the QC. We adjusted all the models for sex and 10 first genetic principal components (PCs). We run two different tests: a single test and a group test per gene.

## Single test

The single test is a classical GWAS, where all the variants are tested against the phenotype. We used a Wald test to assess the associations significance.

```{r}
lambdaClayton <- function(x, trim = 0.5) {
    xx <- qnorm(1-x)^2
    N <- length(xx)
    obsvd <- sort(xx, na.last=NA)
    expctd <- qchisq(p = (1:N)/(N + 1), 1)
    Nu <- floor(trim * N)
    lambda <- mean(obsvd[1:Nu])/mean(expctd[1:Nu])
    lambda
}
```

```{r}
#single <- read_delim("../results/associations/GWAS/2021-04-14/Obesity.GWAS.plink.assoc.logistic",
 #                    progress = TRUE, col_names = TRUE, delim = "\t")
# single$CHROM[is.na(single$CHROM)] <- 23
# single.snp <- subset(single, !Test %in% c(LETTERS[1:10], "Sex"))
# single.snp <- subset(single.snp, !is.na(Pvalue))
single <- read_table2("../results/associations/GWAS/2021-04-15/Obesity.GWAS.plink.assoc.logistic",
                     col_names = TRUE)

single.snp <- subset(single, TEST == "ADD" & !is.na(P))
```

```{r}
qq(single.snp$P)
lambdaClayton(single.snp$P)
```

We do not observe inflation in the GWAS results

```{r}
manhattan(single.snp)
```

We observed `r sum(single.snp$P < 5e-8)` SNPs with genome-wide significant association (P < 5e-8) with the extreme obesity, distributed in chromosomes 1, 6, 7, 12, 16, 17, 18, 19, 22 and 23.

We run two approaches to define loci:

- Clumping with plink: plink generated loci starting by SNPs with genome-wide significance (P < 5.e-8). The loci were extended to include all nearby SNPs with P < 1e-5 and in high LD.
- Depict: depict based on SNPs with P < 1e-5. Performs a similar clustering but also includes nearby genes.

```{r}
lociplink <- read_table2("../results/associations/GWAS/2021-04-15/Obesity.GWAS.plink.clumped",
                     progress = TRUE, col_names = TRUE)
locidepict <- read_delim("../results/associations/GWAS/2021-04-15/assoc_loci.txt",
                     progress = TRUE, col_names = TRUE, comment = "#", delim = "\t")

## Create range with SNPs in SP2
getNum <- function(SP2){
    if (SP2 == "NONE"){
        return(NA)
    }
    a <- strsplit(SP2, ":")[[1]]
    as.numeric(a[seq(2, length(a), 3)])
}
    
lociplink.range <- lociplink %>%
    mutate(minClump = sapply(SP2, function(x) min(getNum(x))),
           maxClump = sapply(SP2, function(x) max(getNum(x))),
           start = pmin(minClump, BP, na.rm = TRUE), 
           end = pmax(minClump, BP, na.rm = TRUE)) 
lociplink.range$id <- paste0("plink", seq_len(nrow(lociplink.range)))
plinkGR <- makeGRangesFromDataFrame(lociplink.range, start.field = "BP", end.field = "BP")
plinkGR$id <- paste0("plink", seq_len(length(plinkGR)))

locidepict$id <- paste0("depict", seq_len(nrow(locidepict)))
depictGR <- makeGRangesFromDataFrame(locidepict)
depictGR$id <- paste0("depict", seq_len(length(depictGR)))
ov <- findOverlaps(plinkGR, depictGR)
plink.un <- lociplink.range[-from(ov), ]
comGR <- c(plinkGR[-from(ov)], depictGR[unique(to(ov))])
```

Overall, we found `r length(plinkGR)` loci with plink and `r length(depictGR)` loci with depict. After combining both sets, and discarding depict loci without genome-wide significant SNPs, we have a total  `r length(comGR)` loci with genome-wide significance.

Next, we removed those loci owith an overlap > 30% with segmental duplications. 

```{r}
LCRs <- read.table("/home/SHARED/DATA/REFERENCES/GRCh37/SegmentalDups/RLCRs_chrNCBI_no_Repeat_Masker.txt", comment.char = "", header = FALSE, as.is = TRUE)
LCRGR <- makeGRangesFromDataFrame(LCRs, seqnames.field = "V1",
                                  start.field = "V2", end.field = "V3")

over.LCR <- findOverlaps(comGR, LCRGR)
comGR.filt <- comGR[-unique(from(over.LCR))]
good <- sapply(unique(from(over.LCR)), function(i){
    
    segDup.sel <-  LCRGR[to(ov)[from(ov) == i]]
    sum(width( segDup.sel)) < 0.3*width(comGR[i])
})
comGR.filt <- c(comGR.filt, comGR[unique(from(over.LCR))][good])
```

We got a final set of `r length(comGR.filt)` loci with significant genome-wide association with extreme obesity and outside segmental duplications regions. 

```{r, cache = TRUE}
options(timeout = 3600)
curgwas <- makeCurrentGwascat(genome = "GRCh37")
bmi.gwas <- subsetByTraits(curgwas, tr = c("Body mass index", "Obesity and osteoporosis",
                                           "Overweight status",
                                           "Birth weight", 
                                           "Underweight status",
                                           "Birth weight variance",
                                           "Offspring birth weight",
                                           "Obesity", "Obesity (extreme)",
                                           "Obesity (early onset extreme)", 
                                           "Obesity-related traits", "Childhood obesity"))
```
Next, we checked the overlap with hits associated with obesity or body mass index in the GWAS catalog. We used a window of 500Kb to identify matches with previous loci. 

```{r}
ov.nogwas <- findOverlaps(comGR.filt + 500e3, bmi.gwas)
comGR.nogwas <- comGR.filt[-unique(from(ov.nogwas))]
```

`r length(unique(from(ov.nogwas)))` loci were previously described in GWAS catalog while `r length(comGR.nogwas)` loci are novel. We will explore these loci with further detail. 

```{r}
plink.new <- subset(lociplink.range, id %in% comGR.nogwas$id)
depict.new <- subset(locidepict, id %in% comGR.nogwas$id)

```

## Inversions

We associated the chromosomal inversions from scoreInvHap with extreme obesity. We adjusted all the models by the first 10 principal components of genetic variance. 

```{r}
load("../results/associations/inversions/2021-04-16/Obesity.inversions.assoc.Rdata")
inv_assocs
```
Only inv16_009 was associated with extreme obesity.

```{r}
attr(inv_assocs, "tables")$inv16_009
```

Inverted allele is more frequent in case samples. The results support a dominant model, where having at least one inverted allele have 1.41 more changes of having obesity. 


# Genetic scores

We obtained polygenic risk scores from a model of BMI obtained from GWAS catalog (PGS000027).

```{r}
iniQCfam <- dir(path = "../results/preImputation/plink/", pattern = "fam",
  full.names = TRUE, recursive = TRUE)
names(iniQCfam) <- gsub(".fam", "", basename(iniQCfam))

famObj <- lapply(iniQCfam, read.delim, header = FALSE, sep = " ")
famDf <- Reduce(rbind, famObj) %>%
  mutate(Dataset = rep(names(iniQCfam), sapply(famObj, nrow)),
         IID = V2) %>%
  dplyr::select(IID, Dataset)

pcs <- read.table("../results/postImputation/stats/2021-04-06/mergedDataset.filtered.eigenvec")
colnames(pcs) <- c("FID", "IID", paste0("PC", 1:10))

pgs <- read.table("../results/associations/PRS/2021-04-16/obesity.profile", 
                  header = TRUE) %>%
    mutate(STATUS = ifelse(grepl("SB", IID), "Control", "Case"),
           STATUS = factor(STATUS, levels = c("Control", "Case")),
           PGS = (SCORESUM - mean(SCORESUM))/sd(SCORESUM), 
           PGS_quantile = cut(PGS, 
                              breaks = quantile(PGS, seq(0, 1, 0.1)),
                              labels = paste0("Q", 1:10),
                              include.lowest = TRUE),
           PGS_carrier = ifelse(PGS_quantile == "Q10", "Carrier", "Reference"),
           PGS_carrier = factor(PGS_carrier, levels = c("Reference", "Carrier"))) %>%
    left_join(dplyr::select(pcs, -FID), by = "IID") %>%
    left_join(famDf, by = "IID") %>%
    as_tibble()

```
```{r}
pgs %>%
    ggplot(aes(x = PGS, color = STATUS)) +
    geom_density() +
    theme_bw() +
    geom_vline(xintercept = 0)

pgs %>% 
    lm(PGS ~ STATUS, .) %>%
    summary()

pgs %>% 
    dplyr::select(starts_with("PC"), STATUS, PGS) %>%
    lm(PGS ~ ., .) %>%
    summary()
```

Case samples have higher polygenic risk scores than control samples. This difference is significant even after adjusting for genetic principal components. 


```{r}
pgs_Qtab <- pgs %>%
    mutate(PGS_cat = ifelse(PGS_quantile == "Q1", "Bottom quantile", 
                            ifelse(PGS_quantile == "Q10", "Top quantile", 
                                   "Quantiles 2-9")),
           PGS_cat = factor(PGS_cat, 
                            levels = c("Bottom quantile", "Quantiles 2-9", "Top quantile"))) %>%
    group_by(STATUS, PGS_cat) %>%
    summarize(n = n()) %>%
    group_by(PGS_cat) %>%
    mutate(prop = n/sum(n))
pgs_Qtab %>%
    ggplot(aes(x = PGS_cat, y = prop, fill = PGS_cat)) +
    geom_bar(stat = "identity", colour = "black") +
    facet_grid(~ STATUS) +
    xlab("Disease Status") +
    ylab("Percent") +
    scale_fill_manual(name = "Polygenic score", values = c("#EEEEEE", "#999999", "#444444"))
```

The proportion of individuals with extreme obesity is much higher in the top quantile than in the bottom quantile.

```{r}
pgs %>% 
    dplyr::select(starts_with("PC"), STATUS, PGS_carrier) %>%
    glm(STATUS ~ ., ., family = "binomial") %>%
    summary()

pgs %>% 
    dplyr::select(starts_with("PC"), STATUS, PGS_carrier) %>%
    glm(STATUS ~ ., ., family = "binomial") %>%
    logistic.display()
```

Individuals in the top decile of the PRS have an OR of 3.3 of being extreme obese.

## Stratification by cohort

We examined whether we could find differences between the cohorts, which can be caused by technical differences during implementation.

```{r}
pgs %>%
    ggplot(aes(x = PGS, color = Dataset)) +
    geom_density() +
    theme_bw() +
    geom_vline(xintercept = 0)

pgs %>% 
    dplyr::select(starts_with("PC"), Dataset, PGS) %>%
    lm(PGS ~ ., .) %>%
    summary()

pgs %>% 
    dplyr::select(starts_with("PC"), Dataset, PGS) %>%
    mutate(Dataset = factor(Dataset, levels = c("Obesity2", "Obesity1", "Obesity3", "Epicuro"))) %>%
    lm(PGS ~ ., .) %>%
    summary()
```

There are no significant differences in mean PGS between the three cohorts.